rfd/XXXX-cloud-agent-upgrades.md

```diff
@@ -0,0 +1,377 @@
+
+
+## What
+
+A simple mechanism for automatically keeping cloud agents on a common "stable"
+install channel.
+
+
+## Why
+
+We would like to increase the cadence at which cloud clusters can be upgraded, and
+eliminate the significant overhead incurred by requiring users to manually upgrade agents.
+
+The existing upgrade system proposal is very complex, and many of its features
+are only relevant for niche on-prem scenarios. The majority of cloud deployments can be served
+by a far simpler system.
+
+This proposal describes a simple framework for keeping cloud agents up to date, designed with the
+intent to rely as little as possible on teleport itself for the upgrades.
+
+Because we are striving for as simple a system as possible, we will make a few scope-limiting
+assumptions/concessions:
+
+- All enrolled cloud agents track a single global cloud-stable channel, rather than having a
+per-cluster target agent version.
+
+- Only a limited subset of common deployment scenarios will be supported.
+
+- The only supported form of "rollback" will be to downgrade the global cloud-stable target itself.
+
+- We won't be trying to come up with a solution that works for everyone. Some cloud clusters will have
+this behavior enabled, and some won't.
+
+
+## Details
+
+From the start, we will support the following scenarios:
+
+- Teleport agents deployed as `k8s` pods using a helm chart the explicitly supports this upgrade system.
+
+- Teleport agents installed via a dedicated `apt` or `yum` package that sets up the installer.
+
+Agents deployed in one of the above manners will automically be "enrolled" in the cloud upgrade system as part of the
+installation process. Enrollment will not be coordinated at the cluster-level (i.e. agents will not check with the
+cluster to determine if upgrades should occur). It will be assumed that the latest available cloud-stable package is
+always the "correct" package to be running. This is an important divergence from how teleport versioning is typically
+handled, and has the effect of allowing unhealthy teleport instances to attempt upgrading as a means of "fixing" themselves.
+
+The only form of cluster-level coordination around agent upgrades will be for timing. Since a restart is
+necessary for an upgrade to take effect, auth healthy agents will periodically load and store an agent maintenance
+window/schedule, and seek to update only during a valid window. Unhealthy agents will be permitted to be
+upgraded outside of the maintenance window in an attempt to become healthy.
+
+Teleport cloud clusters that track cloud-stable will have their control planes frequently upgraded s.t.
+their control planes are always compatible with the current version targeted by the `cloud-stable` channel.
+
+The current `cloud-stable` target version will be served via an `s3` bucket backed by CloudFront. Ex:
+
+```
+$ VERSION="$(curl --proto '=https' --tlsv1.2 -sSf https://update.gravitational.io/v1/cloud-stable/version)"
```

Comment:
Our repos are on https://charts.releases.teleport.dev/ and are managed with infrastructure as a code. I think we should be using `updates.releases.teleport.dev` or similar subdomain of `teleport.dev`. I think we are backed by cloudfront + S3 as well.

Reply:
@fheinecke does the above suggestion jive with your plans?

Reply:
> Are we using gravitational.io for anything today? What teleport agents are dialing to check for updates?

The use of `update.gravitational.io` is just a placeholder.  I believe @fheinecke has been chatting with cloud/release-service folks to determine where to actually put the cloud-stable endpoint.

Reply:
Sure that works for me. As @fspmarshall said `update.gravitational.io` was just a placeholder for discussion. I'd like to talk with release service folks to determine if a separate domain/subdomain (vs path under the release service) would be preferable, but ultimately it doesn't make a huge difference.